add an intro to Ember Data #164

jenweber opened the pull request:
Part of #164 

Please review for clarity for new users.

jenweber commented:
@runspired would you take a look?

CodingItWrong commented:
Great, simple intro! I left a few miscellaneous suggestions and questions

jenweber commented on guides/v3.4.0/models/index.md:
```diff
@@ -1,5 +1,18 @@
-Models are objects that represent the underlying data that your
-application presents to the user. Different apps will have very
+Ember developers have great options for how they handle data from back end APIs. Ember itself works with almost any back end, whether it follows general REST, JSONAPI, GraphQL, ElasticSearch, or none of the above.  
```
I like the shortest version a lot. Done.

jenweber commented on guides/v3.4.0/models/index.md:
```diff
@@ -1,5 +1,18 @@
+Many developers choose to use Ember Data, a powerful set of tools for formatting requests, serializing responses, and efficiently managing a local cache of data. 
+The Ember Data library is included by default in Ember apps. 
+Alternately, some developers write all their own code to handle API requests, using native JavaScript methods like [fetch](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) or third party libraries.
+Of course, many apps use a combination of approaches.
```
good catch, done

jenweber commented on guides/v3.4.0/models/index.md:
```diff
@@ -1,5 +1,18 @@
+
+## What are Ember Data Models?
+
+In Ember Data, Models are objects that represent the underlying data that your
```
Yeah, you're right, this should be consistent. Done.

jenweber commented on guides/v3.4.0/models/index.md:
```diff
@@ -1,5 +1,18 @@
+Alternately, some developers write all their own code to handle API requests, using native JavaScript methods like [fetch](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) or third party libraries.
+Of course, many apps use a combination of approaches.
+
+This section of the Guides describes the essential features of Ember Data. To learn about other ways to handle data in an app, check out [Ember Observer](https://www.emberobserver.com/) for addons or search for community-made tutorials.
```
I am on the fence and err on the side of the more generic link. The reasoning is, if I start thinking things like "hmm, is there anything for firebase? What about Rails?" and then I look at the `data` tag, I might think there's nothing available. Whereas, if we link to main Ember Observer, people get the idea that they should search/explore. I guess I don't want it to look like the `data` category is exhaustive.

jenweber commented on guides/v3.4.0/models/index.md:
```diff
@@ -1,5 +1,18 @@
+
+Many developers choose to use Ember Data, a powerful set of tools for formatting requests, serializing responses, and efficiently managing a local cache of data. 
+The Ember Data library is included by default in Ember apps. 
+Alternately, some developers write all their own code to handle API requests, using native JavaScript methods like [fetch](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) or third party libraries.
```
yes, done

ctcpip commented on guides/v3.4.0/models/index.md:
```diff
@@ -1,5 +1,19 @@
+Alternately, some developers write all their own code to handle API requests, using native JavaScript methods like [fetch](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) or third-party libraries.
+Many apps use a combination of approaches.
+
+This section of the Guides describes the essential features of Ember Data. To learn about other ways to handle data in an app, check out [Ember Observer](https://www.emberobserver.com/) for addons or search for community-made tutorials.
```
I'm not sure suggesting a visit to ember observer should be the first recommendation to someone interested in alternatives to using `ember-data`.  We may not want to reference specific 3rd party tutorials, either though.  Maybe just remove the whole 'to learn about other ways...' sentence?

jenweber replied:
I think it’s important to show that people have options. Sometimes people abandon ember because they think they don’t, and other times people think they need to write everything themselves. I want to present the realistic range, while still encouraging the use of Ember Data. I will reword some.

ctcpip commented on guides/v3.4.0/models/index.md:
```diff
@@ -1,5 +1,19 @@
+
+In Ember Data, models are objects that represent the underlying data that your
+application presents to the user. 
+(Note that Ember Data models are a different concept than the `model` hook method on Routes, although they share the same name.)
```
Although generally speaking, isn't the purpose of the `model` hook to load the ember-data `model`(s) for presentation?  The note seems to suggest they are not related, even though it is fair to say they are linked ("Typically, the model hook should return an Ember Data record" from route model guide).  Also, in the route model guide, when it references the ember-data record, it links to this page, so maybe this parenthetical is unnecessary.

ctcpip replied:
Yet that is what the guide says - "Typically, the model hook should return an Ember Data record"

jenweber replied:
“Typically, the model hook should return an Ember Data record, but it can also return any promise object (Ember Data records are promises), or a plain JavaScript object or array.“ [link](
https://guides.emberjs.com/release/routing/specifying-a-routes-model/)

Is there a way the quote above from the models hook guide can be improved?

Given the amount of confusion that is common with new devs, I’m ok with repeating the point. I’ll read over the model hook docs and open issues for possible improvements.

However I should probably link to the model hook info here.

jenweber replied:
I have removed the parens, added a link, and used the word “different” concept instead of “separate”

ctcpip commented on guides/v3.4.0/models/index.md:
```diff
@@ -1,5 +1,19 @@
+Ember itself works with any type of back end: REST, JSON API, GraphQL, or anything else.
+
+Many developers choose to use Ember Data, a powerful set of tools for formatting requests, serializing responses, and efficiently managing a local cache of data. 
+The Ember Data library is included by default in Ember apps. 
```
this sounds a lot like the paragraph later in the document:

> Ember Data, included by default when you create a new application, is a
>  library that integrates tightly with Ember to make it easy to retrieve
>  models from your server as JSON, save updates back to the server, and
>  create new models in the browser.

jenweber replied:
Good catch. I removed the redundant description.

runspired commented on guides/v3.4.0/models/index.md:
```diff
@@ -1,5 +1,19 @@
+Ember itself works with any type of back end: REST, JSON API, GraphQL, or anything else.
+
+Many developers choose to use Ember Data, a powerful set of tools for formatting requests, serializing responses, and efficiently managing a local cache of data. 
+The Ember Data library is included by default in Ember apps. 
```
Perhaps:

> `ember-data` is included by default for applications generated with `ember-cli`; however, if you do not wish to use `ember-data` it can easily be removed by removing the `ember-data` entry from `package.json`.

jenweber replied:
Seems good! Done.

runspired commented on guides/v3.4.0/models/index.md:
```diff
@@ -1,5 +1,19 @@
+Ember developers have great options for how they handle data from back end APIs. 
+Ember itself works with any type of back end: REST, JSON API, GraphQL, or anything else.
+
+Many developers choose to use Ember Data, a powerful set of tools for formatting requests, serializing responses, and efficiently managing a local cache of data. 
```
I will bring this up with the `ember-data` team, but I think it is time we officially settled on how to reference the library in technical writing. Some places we use `Ember Data`, some places `ember-data`, and still other places `EmberData`. I am preferential to using `ember-data` as it matches the package name.

locks replied:
We have styled it "Ember Data", "Ember", and "Ember CLI" throughout our resources, unless specifically mentioning the repository/library, like "it can easily be removed by removing the `ember-data` entry in `package.json`." mentioned below.

runspired replied:
It would be nice to champion a change away from `Ember Data` to `ember-data`. Unfortunately, I too often see users incorrectly confusing/mixing concepts from `Ember` and `Ember Data` and a good deal of it seems to stem from the branding.

jenweber replied:
I have opened an issue for discussion from the learning team perspective, #184.

I think calling it ember-data has some nice benefits. I will leave it as Ember Data, for the scope of this PR, but please let me know what the data team settles on.

runspired commented on guides/v3.4.0/models/index.md:
```diff
@@ -1,5 +1,19 @@
+Ember developers have great options for how they handle data from back end APIs. 
+Ember itself works with any type of back end: REST, JSON API, GraphQL, or anything else.
+
+Many developers choose to use Ember Data, a powerful set of tools for formatting requests, serializing responses, and efficiently managing a local cache of data. 
```
> for formatting requests, serializing responses, and efficiently managing a local cache of data. 

Typically outbound (from `ember-data`'s store) is "serialization" and inbound (to `ember-data`'s store) is "normalization". Alternatives to "normalization" include `deserialization` or language around `munging` `aligning` or `formatting` responses into `JSON API`.

jenweber replied:
Fixed. I'll call it normalization here. I think that term has enough traction outside of Ember Data.

runspired commented on guides/v3.4.0/models/index.md:
```diff
@@ -1,5 +1,19 @@
+## What are Ember Data models?
+
+In Ember Data, models are objects that represent the underlying data that your
+application presents to the user. 
```
I am unsure what the best language here is, but here is how we use various terms on the `ember-data` team:

- `Model` => the `class` definition containing schema information about what `attributes` and `relationships` are valid for a particular JSON API `resource` whose `type` matches the `modelName` (the filepath).
- `record` => an instance of a `Model` for presenting the data of a single `resource` to the UI

jenweber commented:
Hey @runspired @ctcpip and @CodingItWrong, I think I've addressed all your comments! Let me know if there are any other changes to make before we merge. Thanks again for your help! It's part of people's first impressions of Ember Data in the Guides, so I want to get it right.

jenweber commented:
If there are no objections, I think this might be good to merge now.

ctcpip commented on guides/v3.4.0/models/index.md:
```diff
@@ -1,7 +1,33 @@
+[`model`](../routing/specifying-a-routes-model/) method on Routes, 
+although they share the same name. 
+
+Different apps will have very
```
maybe to strong to say "will" have ?  Perhaps change to "may" have
